includes/derivatives.inc

```diff
@@ -237,23 +237,47 @@ function islandora_ocr_get_converted_source(AbstractDatastream $datastream) {
 }
 
 /**
- * Derive the OCR stream from OBJ.
+ * Derives an 'OCR' or 'HOCR' datastream.
+ *
+ * @param string $type
+ *   Either "OCR" or "HOCR".
+ * @param string $rels_flag
+ *   String literal flag in the RELS-EXT to use to determine whether or not we
+ *   should proceed with derivative generation.
+ * @param AbstractObject $object
+ *   The object on which to derive a datastream.
+ * @param bool $force
+ *   Whether to force derivative generation.
+ *
+ * @throws InvalidArgumentException
+ *   If the $type is neither OCR nor HOCR.
+ *
+ * @return array
+ *   A derivative success/fail structured array.
  */
-function islandora_ocr_derive_ocr(AbstractObject $object, $force = FALSE) {
+function islandora_ocr_derive_datastream($type, $rels_flag, AbstractObject $object, $force = FALSE) {
```

Comment:
`$type` and `$rels_flag` are co-dependent, yeah? Probably better to eliminate one and look up the other... Possibly build up an array with the relevant info to use later?... Something like:

```php
function islandora_ocr_derive_datastream($type, AbstractObject $object, $force = FALSE) {
  $map = array(
    'HOCR' => array(
      'function' => 'islandora_ocr_create_hocr',
      'flag' => ISLANDORA_OCR_GENERATE_HOCR_RELS_STATEMENT,
    ),
    'OCR' => array(
      'functions' => 'islandora_ocr_create_ocr',
      'flag' => ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT,
    ),
  );

  if ($updated_date = variable_get(ISLANDORA_OCR_UPDATE_7001_TIMESTAMP, FALSE)) {
    // Copy ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT to ISLANDORA_OCR_GENERATE_HOCR_RELS_STATEMENT for objects that still have the flag and were created when only the former existed.
    // ... Could possibly isolate this to a separate function, just passing in the object?
    $diff = date_diff(new DateTime($timestamp), $object->createdDate);
    // Negative diffs indicate an object created before the timestamp.
    if ($diff->invert) {
      $rels = $object->relationships;
      $initial_autocommit = $rels->autoCommit;
      foreach ($rels->get(ISLANDORA_RELS_EXT_URI, ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT)) as $result) {
        $rels->autoCommit = FALSE;
        $rels->add(ISLANDORA_RELS_EXT_URI, ISLANDORA_OCR_GENERATE_OCR_RELS_STATEMENT, $result['object']['value'], RELS_TYPE_PLAIN_LITERAL);
      }
      $rels->autoCommit = $initial_autocommit;
    }
  }

  if (!isset($map[$type)) {
    throw new InvalidArgumentException(t('Invalid OCR datastream type "@type"', array(
      '@type' => $type,
    )));
  }
  $rels_flag = $map[$type]['flag'];
  $function = $map[$type]['function'];

  [...]
  
  $derivative_file = $function($file, array('language' => $language));

  [...]

  _islandora_ocr_remove_generation_rels_ext_statement($object, $rels_flag);
}

[...]

function _islandora_ocr_remove_generation_rels_ext_statement(AbstractObject $object, $flag) {
  $rels = $object->relationships;
  if ($rels->get(ISLANDORA_RELS_EXT_URI, $flag, NULL, RELS_TYPE_PLAIN_LITERAL)) {
    $rels->remove(ISLANDORA_RELS_EXT_URI, $flag, NULL, RELS_TYPE_PLAIN_LITERAL);
  }
}
```